File: 3414.maximum-score-of-non-overlapping-intervals.py
#
# @lc app=leetcode id=3414 lang=python3
#
# [3414] Maximum Score of Non-overlapping Intervals
#

# @lc code=start
from typing import List

class Solution:
    def maximumWeight(self, intervals: List[List[int]]) -> List[int]:
        n = len(intervals)
        if n == 0:
            return []
        events = [(inter[0], inter[1], inter[2], i) for i, inter in enumerate(intervals)]
        events.sort(key=lambda x: x[1])
        K = 4
        prefix_max = [[0] * (K + 1) for _ in range(n + 1)]
        prefix_paths = [[[] for _ in range(K + 1)] for _ in range(n + 1)]
        for i in range(n):
            start, end_time, weight, idx = events[i]
            # Compute dp for ending at i
            dp_1 = weight
            path_1 = [idx]
            # Binary search for largest m < i s.t. events[m][1] < start
            left, right = 0, i - 1
            m = -1
            while left <= right:
                mid = left + (right - left) // 2
                if events[mid][1] < start:
                    m = mid
                    left = mid + 1
                else:
                    right = mid - 1
            temp_dp = [0] * (K + 1)
            temp_paths = [[] for _ in range(K + 1)]
            temp_dp[1] = dp_1
            temp_paths[1] = path_1
            for j in range(2, K + 1):
                if m < 0:
                    break
                prev_max_s = prefix_max[j - 1][m + 1]
                if prev_max_s > 0:
                    new_score = prev_max_s + weight
                    temp_dp[j] = new_score
                    prev_path = prefix_paths[j - 1][m + 1][:]
                    new_path = sorted(prev_path + [idx])
                    temp_paths[j] = new_path
            # Update prefix for i+1: max(skip i, take i)
            for j in range(1, K + 1):
                score_skip = prefix_max[j][i]
                path_skip = prefix_paths[j][i]
                score_take = temp_dp[j]
                path_take = temp_paths[j]
                if score_take > score_skip:
                    prefix_max[j][i + 1] = score_take
                    prefix_paths[j][i + 1] = path_take
                elif score_take == score_skip:
                    prefix_max[j][i + 1] = score_skip
                    if path_take < path_skip:
                        prefix_paths[j][i + 1] = path_take
                    else:
                        prefix_paths[j][i + 1] = path_skip
                else:
                    prefix_max[j][i + 1] = score_skip
                    prefix_paths[j][i + 1] = path_skip
        # Global max score and lex min path
        max_score = 0
        best_list = []
        for j in range(1, K + 1):
            curr_score = prefix_max[j][n]
            curr_path = prefix_paths[j][n]
            if curr_score > max_score or (curr_score == max_score and (not best_list or curr_path < best_list)):
                max_score = curr_score
                best_list = curr_path
        return best_list

# @lc code=end

File: 3414.maximum-score-of-non-overlapping-intervals.java
//
// @lc app=leetcode id=3414 lang=java
//
// [3414] Maximum Score of Non-overlapping Intervals
//
// @lc code=start
class Solution {
    public int[] maximumWeight(List<List<Integer>> intervals) {
        int n = intervals.size();
        int[][] arr = new int[n][4]; // [l, r, weight, original_index]
        for (int i = 0; i < n; ++i) {
            List<Integer> iv = intervals.get(i);
            arr[i][0] = iv.get(0);
            arr[i][1] = iv.get(1);
            arr[i][2] = iv.get(2);
            arr[i][3] = i;
        }
        java.util.Arrays.sort(arr, java.util.Comparator.comparingInt(a -> a[1]));
        int[] ends = new int[n];
        for (int i = 0; i < n; ++i) ends[i] = arr[i][1];
        int[] prev = new int[n];
        for (int i = 0; i < n; ++i) {
            int l = arr[i][0];
            int lo = 0, hi = i - 1, res = -1;
            while (lo <= hi) {
                int mid = (lo + hi) / 2;
                if (ends[mid] < l) {
                    res = mid;
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
            prev[i] = res;
        }
        int K = 4;
        long[][] dp = new long[n + 1][K + 1];
        int[][][] idx = new int[n + 1][K + 1][K];
        for (int i = 0; i <= n; ++i)
            for (int k = 0; k <= K; ++k)
                for (int j = 0; j < K; ++j)
                    idx[i][k][j] = -1;
        for (int i = 1; i <= n; ++i) {
            for (int k = 1; k <= K; ++k) {
                // Option 1: not take current
                dp[i][k] = dp[i - 1][k];
                System.arraycopy(idx[i - 1][k], 0, idx[i][k], 0, K);
                // Option 2: take current
                int j = prev[i - 1] + 1;
                long candidate = dp[j][k - 1] + arr[i - 1][2];
                if (candidate > dp[i][k] || (candidate == dp[i][k] && lexSmaller(idx[j][k - 1], k - 1, idx[i][k], k, arr[i - 1][3]))) {
                    dp[i][k] = candidate;
                    for (int m = 0; m < k - 1; ++m) idx[i][k][m] = idx[j][k - 1][m];
                    idx[i][k][k - 1] = arr[i - 1][3];
                    for (int m = k; m < K; ++m) idx[i][k][m] = -1;
                }
            }
            // k=0
            dp[i][0] = dp[i - 1][0];
            System.arraycopy(idx[i - 1][0], 0, idx[i][0], 0, K);
        }
        long best = -1;
        int[] ans = new int[0];
        for (int k = 1; k <= K; ++k) {
            if (dp[n][k] > best) {
                best = dp[n][k];
                ans = extract(idx[n][k], k);
            } else if (dp[n][k] == best && lexSmaller(idx[n][k], k, ans, ans.length, -1)) {
                ans = extract(idx[n][k], k);
            }
        }
        return ans;
    }
    private boolean lexSmaller(int[] a, int lenA, int[] b, int lenB, int extra) {
        int[] A = new int[lenA + (extra >= 0 ? 1 : 0)];
        for (int i = 0; i < lenA; ++i) A[i] = a[i];
        if (extra >= 0) A[lenA] = extra;
        java.util.Arrays.sort(A);
        int[] B = new int[lenB];
        for (int i = 0; i < lenB; ++i) B[i] = b[i];
        java.util.Arrays.sort(B);
        int n = Math.min(A.length, B.length);
        for (int i = 0; i < n; ++i) {
            if (A[i] < B[i]) return true;
            if (A[i] > B[i]) return false;
        }
        return A.length < B.length;
    }
    private int[] extract(int[] arr, int k) {
        int[] res = new int[k];
        for (int i = 0; i < k; ++i) res[i] = arr[i];
        java.util.Arrays.sort(res);
        return res;
    }
}
// @lc code=end